# Practice B. Normalize the data
- Repeat part A but use a normalized version of the data.
#####  Dataset
- [Download dataset here](https://cocl.us/concrete_data)

In [1]:
# Import packages
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

import keras
from keras.models import Sequential
from keras.layers import Dense

from sklearn.metrics import mean_squared_error
from sklearn.model_selection import train_test_split

# Read in the data
concrete_data = pd.read_csv("/Users/pc/Desktop/IBM AI Engineer/Introduction to Deep Learning & Neural Networks with Keras/concrete_data.csv")

# Split data into predictors and target
concrete_data_columns = concrete_data.columns
predictors = concrete_data[concrete_data_columns[concrete_data_columns != "Strength"]]
target = concrete_data["Strength"]

2023-08-16 15:06:57.642956: I tensorflow/core/platform/cpu_feature_guard.cc:182] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.


### Normalize the data

In [2]:
# Normalization of the data
predictors_norm = (predictors - predictors.mean()) / predictors.std()
target = (target - target.mean()) / target.std()

# the number of predictors
n_cols = predictors_norm.shape[1]

In [3]:
# define the regression model
def regression_model():
    # create model
    model = Sequential()
    model.add(Dense(10, activation = "relu", input_shape = (n_cols, )))
    model.add(Dense(1))
    
    # compile model
    model.compile(optimizer = "adam", loss = "mean_squared_error")
    return model

# build the model
model = regression_model()

MSE_list = []
trial_times = 50
for t in range(trial_times):
    # train / test split
    X_train, X_test, y_train, y_test = train_test_split(predictors, 
                                                    target, 
                                                    test_size=0.3, 
                                                    random_state=42)
    # fit the model
    model.fit(X_train, y_train, epochs = 50, verbose = 2)
    
    # predict values
    predictions = model.predict(X_test)

    # Mean Squared Errors between predictions and the actual values: y_test
    predictions = predictions.reshape(predictions.size, )
    MSE_errors = mean_squared_error(y_test, predictions)
    MSE_list.append(MSE_errors)
    #print("The MSE errors at the {} epoch is {}.".format(e + 1, MSE_errors))


Epoch 1/50
23/23 - 0s - loss: 1013.7368 - 438ms/epoch - 19ms/step
Epoch 2/50
23/23 - 0s - loss: 782.9404 - 19ms/epoch - 814us/step
Epoch 3/50
23/23 - 0s - loss: 604.7643 - 19ms/epoch - 815us/step
Epoch 4/50
23/23 - 0s - loss: 515.1680 - 19ms/epoch - 810us/step
Epoch 5/50
23/23 - 0s - loss: 435.8297 - 19ms/epoch - 816us/step
Epoch 6/50
23/23 - 0s - loss: 376.5018 - 18ms/epoch - 795us/step
Epoch 7/50
23/23 - 0s - loss: 318.8156 - 18ms/epoch - 770us/step
Epoch 8/50
23/23 - 0s - loss: 269.8211 - 19ms/epoch - 807us/step
Epoch 9/50
23/23 - 0s - loss: 235.2489 - 18ms/epoch - 761us/step
Epoch 10/50
23/23 - 0s - loss: 206.0535 - 18ms/epoch - 771us/step
Epoch 11/50
23/23 - 0s - loss: 180.5647 - 18ms/epoch - 796us/step
Epoch 12/50
23/23 - 0s - loss: 160.6783 - 18ms/epoch - 765us/step
Epoch 13/50
23/23 - 0s - loss: 144.8755 - 18ms/epoch - 776us/step
Epoch 14/50
23/23 - 0s - loss: 129.5552 - 19ms/epoch - 828us/step
Epoch 15/50
23/23 - 0s - loss: 116.0801 - 17ms/epoch - 759us/step
Epoch 16/50
23/23 

23/23 - 0s - loss: 0.6163 - 16ms/epoch - 702us/step
Epoch 28/50
23/23 - 0s - loss: 0.5824 - 16ms/epoch - 708us/step
Epoch 29/50
23/23 - 0s - loss: 0.5721 - 16ms/epoch - 709us/step
Epoch 30/50
23/23 - 0s - loss: 0.6183 - 18ms/epoch - 763us/step
Epoch 31/50
23/23 - 0s - loss: 0.6275 - 18ms/epoch - 769us/step
Epoch 32/50
23/23 - 0s - loss: 0.5520 - 19ms/epoch - 808us/step
Epoch 33/50
23/23 - 0s - loss: 0.5564 - 17ms/epoch - 756us/step
Epoch 34/50
23/23 - 0s - loss: 0.5177 - 17ms/epoch - 733us/step
Epoch 35/50
23/23 - 0s - loss: 0.5206 - 16ms/epoch - 712us/step
Epoch 36/50
23/23 - 0s - loss: 0.5212 - 17ms/epoch - 724us/step
Epoch 37/50
23/23 - 0s - loss: 0.5172 - 16ms/epoch - 689us/step
Epoch 38/50
23/23 - 0s - loss: 0.5362 - 16ms/epoch - 680us/step
Epoch 39/50
23/23 - 0s - loss: 0.5083 - 16ms/epoch - 688us/step
Epoch 40/50
23/23 - 0s - loss: 0.5042 - 16ms/epoch - 680us/step
Epoch 41/50
23/23 - 0s - loss: 0.4924 - 16ms/epoch - 677us/step
Epoch 42/50
23/23 - 0s - loss: 0.5427 - 15ms/epoch -

Epoch 4/50
23/23 - 0s - loss: 0.4810 - 20ms/epoch - 855us/step
Epoch 5/50
23/23 - 0s - loss: 0.5286 - 19ms/epoch - 826us/step
Epoch 6/50
23/23 - 0s - loss: 0.7914 - 18ms/epoch - 803us/step
Epoch 7/50
23/23 - 0s - loss: 0.4416 - 18ms/epoch - 763us/step
Epoch 8/50
23/23 - 0s - loss: 0.5292 - 17ms/epoch - 744us/step
Epoch 9/50
23/23 - 0s - loss: 0.4711 - 18ms/epoch - 803us/step
Epoch 10/50
23/23 - 0s - loss: 0.5015 - 19ms/epoch - 825us/step
Epoch 11/50
23/23 - 0s - loss: 0.5444 - 17ms/epoch - 749us/step
Epoch 12/50
23/23 - 0s - loss: 0.4867 - 17ms/epoch - 756us/step
Epoch 13/50
23/23 - 0s - loss: 0.5748 - 18ms/epoch - 785us/step
Epoch 14/50
23/23 - 0s - loss: 0.4553 - 19ms/epoch - 807us/step
Epoch 15/50
23/23 - 0s - loss: 0.7294 - 17ms/epoch - 750us/step
Epoch 16/50
23/23 - 0s - loss: 0.4233 - 20ms/epoch - 849us/step
Epoch 17/50
23/23 - 0s - loss: 0.4563 - 18ms/epoch - 761us/step
Epoch 18/50
23/23 - 0s - loss: 0.4321 - 18ms/epoch - 787us/step
Epoch 19/50
23/23 - 0s - loss: 0.5014 - 20ms/e

23/23 - 0s - loss: 0.4416 - 16ms/epoch - 689us/step
Epoch 32/50
23/23 - 0s - loss: 0.5321 - 16ms/epoch - 709us/step
Epoch 33/50
23/23 - 0s - loss: 0.5413 - 17ms/epoch - 739us/step
Epoch 34/50
23/23 - 0s - loss: 0.6765 - 22ms/epoch - 940us/step
Epoch 35/50
23/23 - 0s - loss: 0.6353 - 22ms/epoch - 971us/step
Epoch 36/50
23/23 - 0s - loss: 0.5368 - 16ms/epoch - 699us/step
Epoch 37/50
23/23 - 0s - loss: 0.4439 - 17ms/epoch - 748us/step
Epoch 38/50
23/23 - 0s - loss: 0.4315 - 16ms/epoch - 696us/step
Epoch 39/50
23/23 - 0s - loss: 0.4335 - 16ms/epoch - 688us/step
Epoch 40/50
23/23 - 0s - loss: 0.4767 - 16ms/epoch - 706us/step
Epoch 41/50
23/23 - 0s - loss: 0.4402 - 16ms/epoch - 693us/step
Epoch 42/50
23/23 - 0s - loss: 0.6162 - 16ms/epoch - 704us/step
Epoch 43/50
23/23 - 0s - loss: 0.5600 - 23ms/epoch - 1ms/step
Epoch 44/50
23/23 - 0s - loss: 0.4930 - 18ms/epoch - 774us/step
Epoch 45/50
23/23 - 0s - loss: 0.4830 - 16ms/epoch - 712us/step
Epoch 46/50
23/23 - 0s - loss: 0.5536 - 16ms/epoch - 6

23/23 - 0s - loss: 0.4568 - 20ms/epoch - 873us/step
Epoch 8/50
23/23 - 0s - loss: 0.5652 - 17ms/epoch - 754us/step
Epoch 9/50
23/23 - 0s - loss: 0.6141 - 17ms/epoch - 737us/step
Epoch 10/50
23/23 - 0s - loss: 0.7463 - 17ms/epoch - 726us/step
Epoch 11/50
23/23 - 0s - loss: 0.6004 - 16ms/epoch - 689us/step
Epoch 12/50
23/23 - 0s - loss: 0.4416 - 17ms/epoch - 728us/step
Epoch 13/50
23/23 - 0s - loss: 0.4418 - 18ms/epoch - 775us/step
Epoch 14/50
23/23 - 0s - loss: 0.7190 - 16ms/epoch - 709us/step
Epoch 15/50
23/23 - 0s - loss: 0.4549 - 23ms/epoch - 992us/step
Epoch 16/50
23/23 - 0s - loss: 0.5099 - 16ms/epoch - 687us/step
Epoch 17/50
23/23 - 0s - loss: 0.5775 - 15ms/epoch - 673us/step
Epoch 18/50
23/23 - 0s - loss: 0.5095 - 16ms/epoch - 693us/step
Epoch 19/50
23/23 - 0s - loss: 0.5457 - 16ms/epoch - 684us/step
Epoch 20/50
23/23 - 0s - loss: 0.6608 - 16ms/epoch - 686us/step
Epoch 21/50
23/23 - 0s - loss: 0.5114 - 19ms/epoch - 838us/step
Epoch 22/50
23/23 - 0s - loss: 0.4646 - 17ms/epoch - 7

Epoch 34/50
23/23 - 0s - loss: 0.4198 - 19ms/epoch - 825us/step
Epoch 35/50
23/23 - 0s - loss: 0.5067 - 16ms/epoch - 689us/step
Epoch 36/50
23/23 - 0s - loss: 0.5152 - 16ms/epoch - 691us/step
Epoch 37/50
23/23 - 0s - loss: 0.7790 - 16ms/epoch - 714us/step
Epoch 38/50
23/23 - 0s - loss: 0.4412 - 16ms/epoch - 689us/step
Epoch 39/50
23/23 - 0s - loss: 0.5233 - 16ms/epoch - 685us/step
Epoch 40/50
23/23 - 0s - loss: 0.7539 - 16ms/epoch - 677us/step
Epoch 41/50
23/23 - 0s - loss: 0.7632 - 23ms/epoch - 991us/step
Epoch 42/50
23/23 - 0s - loss: 0.4150 - 16ms/epoch - 690us/step
Epoch 43/50
23/23 - 0s - loss: 0.4547 - 16ms/epoch - 688us/step
Epoch 44/50
23/23 - 0s - loss: 0.5021 - 19ms/epoch - 805us/step
Epoch 45/50
23/23 - 0s - loss: 0.5603 - 18ms/epoch - 768us/step
Epoch 46/50
23/23 - 0s - loss: 0.6706 - 16ms/epoch - 695us/step
Epoch 47/50
23/23 - 0s - loss: 0.5700 - 16ms/epoch - 692us/step
Epoch 48/50
23/23 - 0s - loss: 0.4396 - 16ms/epoch - 694us/step
Epoch 49/50
23/23 - 0s - loss: 0.3897 - 

Epoch 10/50
23/23 - 0s - loss: 0.4249 - 17ms/epoch - 753us/step
Epoch 11/50
23/23 - 0s - loss: 0.4690 - 17ms/epoch - 742us/step
Epoch 12/50
23/23 - 0s - loss: 0.4643 - 29ms/epoch - 1ms/step
Epoch 13/50
23/23 - 0s - loss: 0.4611 - 21ms/epoch - 916us/step
Epoch 14/50
23/23 - 0s - loss: 0.6228 - 17ms/epoch - 740us/step
Epoch 15/50
23/23 - 0s - loss: 0.5732 - 17ms/epoch - 732us/step
Epoch 16/50
23/23 - 0s - loss: 0.4889 - 20ms/epoch - 883us/step
Epoch 17/50
23/23 - 0s - loss: 0.4847 - 17ms/epoch - 758us/step
Epoch 18/50
23/23 - 0s - loss: 0.4732 - 17ms/epoch - 728us/step
Epoch 19/50
23/23 - 0s - loss: 0.4675 - 16ms/epoch - 695us/step
Epoch 20/50
23/23 - 0s - loss: 0.4905 - 16ms/epoch - 701us/step
Epoch 21/50
23/23 - 0s - loss: 0.6773 - 15ms/epoch - 671us/step
Epoch 22/50
23/23 - 0s - loss: 0.4350 - 16ms/epoch - 685us/step
Epoch 23/50
23/23 - 0s - loss: 0.4949 - 16ms/epoch - 709us/step
Epoch 24/50
23/23 - 0s - loss: 0.4445 - 16ms/epoch - 686us/step
Epoch 25/50
23/23 - 0s - loss: 0.4708 - 16

Epoch 37/50
23/23 - 0s - loss: 0.5235 - 16ms/epoch - 686us/step
Epoch 38/50
23/23 - 0s - loss: 0.5258 - 16ms/epoch - 690us/step
Epoch 39/50
23/23 - 0s - loss: 0.5505 - 15ms/epoch - 663us/step
Epoch 40/50
23/23 - 0s - loss: 0.5856 - 16ms/epoch - 683us/step
Epoch 41/50
23/23 - 0s - loss: 0.4096 - 16ms/epoch - 684us/step
Epoch 42/50
23/23 - 0s - loss: 0.5750 - 16ms/epoch - 687us/step
Epoch 43/50
23/23 - 0s - loss: 0.4439 - 16ms/epoch - 682us/step
Epoch 44/50
23/23 - 0s - loss: 0.9198 - 16ms/epoch - 680us/step
Epoch 45/50
23/23 - 0s - loss: 0.5578 - 16ms/epoch - 709us/step
Epoch 46/50
23/23 - 0s - loss: 0.4237 - 19ms/epoch - 842us/step
Epoch 47/50
23/23 - 0s - loss: 0.4246 - 18ms/epoch - 766us/step
Epoch 48/50
23/23 - 0s - loss: 0.4231 - 16ms/epoch - 677us/step
Epoch 49/50
23/23 - 0s - loss: 0.4117 - 15ms/epoch - 673us/step
Epoch 50/50
23/23 - 0s - loss: 0.4982 - 15ms/epoch - 662us/step
10/10 [==============================] - 0s 895us/step
Epoch 1/50
23/23 - 0s - loss: 0.5969 - 19ms/epoch

Epoch 13/50
23/23 - 0s - loss: 0.4410 - 16ms/epoch - 698us/step
Epoch 14/50
23/23 - 0s - loss: 0.4893 - 15ms/epoch - 668us/step
Epoch 15/50
23/23 - 0s - loss: 0.4649 - 16ms/epoch - 716us/step
Epoch 16/50
23/23 - 0s - loss: 0.4571 - 16ms/epoch - 684us/step
Epoch 17/50
23/23 - 0s - loss: 0.4672 - 16ms/epoch - 683us/step
Epoch 18/50
23/23 - 0s - loss: 0.4567 - 16ms/epoch - 680us/step
Epoch 19/50
23/23 - 0s - loss: 0.4476 - 15ms/epoch - 674us/step
Epoch 20/50
23/23 - 0s - loss: 0.4330 - 18ms/epoch - 775us/step
Epoch 21/50
23/23 - 0s - loss: 0.4449 - 16ms/epoch - 710us/step
Epoch 22/50
23/23 - 0s - loss: 0.5956 - 16ms/epoch - 680us/step
Epoch 23/50
23/23 - 0s - loss: 0.4089 - 15ms/epoch - 662us/step
Epoch 24/50
23/23 - 0s - loss: 0.5100 - 16ms/epoch - 696us/step
Epoch 25/50
23/23 - 0s - loss: 0.4803 - 16ms/epoch - 677us/step
Epoch 26/50
23/23 - 0s - loss: 0.4659 - 16ms/epoch - 678us/step
Epoch 27/50
23/23 - 0s - loss: 0.4850 - 15ms/epoch - 673us/step
Epoch 28/50
23/23 - 0s - loss: 0.4633 - 

Epoch 40/50
23/23 - 0s - loss: 0.4785 - 16ms/epoch - 687us/step
Epoch 41/50
23/23 - 0s - loss: 0.4364 - 16ms/epoch - 692us/step
Epoch 42/50
23/23 - 0s - loss: 0.4658 - 17ms/epoch - 728us/step
Epoch 43/50
23/23 - 0s - loss: 0.6109 - 17ms/epoch - 722us/step
Epoch 44/50
23/23 - 0s - loss: 0.7011 - 16ms/epoch - 696us/step
Epoch 45/50
23/23 - 0s - loss: 0.6259 - 17ms/epoch - 735us/step
Epoch 46/50
23/23 - 0s - loss: 0.4811 - 16ms/epoch - 691us/step
Epoch 47/50
23/23 - 0s - loss: 0.5555 - 16ms/epoch - 706us/step
Epoch 48/50
23/23 - 0s - loss: 0.6342 - 16ms/epoch - 711us/step
Epoch 49/50
23/23 - 0s - loss: 0.4584 - 16ms/epoch - 714us/step
Epoch 50/50
23/23 - 0s - loss: 0.4923 - 17ms/epoch - 722us/step
10/10 [==============================] - 0s 881us/step
Epoch 1/50
23/23 - 0s - loss: 0.7566 - 19ms/epoch - 816us/step
Epoch 2/50
23/23 - 0s - loss: 0.4274 - 16ms/epoch - 698us/step
Epoch 3/50
23/23 - 0s - loss: 0.4830 - 16ms/epoch - 687us/step
Epoch 4/50
23/23 - 0s - loss: 0.4002 - 22ms/epoch - 

Epoch 16/50
23/23 - 0s - loss: 0.4204 - 16ms/epoch - 710us/step
Epoch 17/50
23/23 - 0s - loss: 0.6012 - 17ms/epoch - 721us/step
Epoch 18/50
23/23 - 0s - loss: 0.7826 - 16ms/epoch - 696us/step
Epoch 19/50
23/23 - 0s - loss: 0.4990 - 16ms/epoch - 715us/step
Epoch 20/50
23/23 - 0s - loss: 0.5659 - 18ms/epoch - 804us/step
Epoch 21/50
23/23 - 0s - loss: 0.5187 - 16ms/epoch - 707us/step
Epoch 22/50
23/23 - 0s - loss: 0.5174 - 16ms/epoch - 697us/step
Epoch 23/50
23/23 - 0s - loss: 0.4377 - 24ms/epoch - 1ms/step
Epoch 24/50
23/23 - 0s - loss: 0.4858 - 17ms/epoch - 747us/step
Epoch 25/50
23/23 - 0s - loss: 0.4890 - 16ms/epoch - 715us/step
Epoch 26/50
23/23 - 0s - loss: 0.5834 - 17ms/epoch - 724us/step
Epoch 27/50
23/23 - 0s - loss: 0.5424 - 16ms/epoch - 714us/step
Epoch 28/50
23/23 - 0s - loss: 0.4719 - 16ms/epoch - 685us/step
Epoch 29/50
23/23 - 0s - loss: 0.3961 - 16ms/epoch - 691us/step
Epoch 30/50
23/23 - 0s - loss: 0.4619 - 16ms/epoch - 697us/step
Epoch 31/50
23/23 - 0s - loss: 0.4365 - 15

Epoch 43/50
23/23 - 0s - loss: 0.4570 - 16ms/epoch - 689us/step
Epoch 44/50
23/23 - 0s - loss: 0.4316 - 16ms/epoch - 692us/step
Epoch 45/50
23/23 - 0s - loss: 0.5319 - 18ms/epoch - 789us/step
Epoch 46/50
23/23 - 0s - loss: 0.5369 - 18ms/epoch - 784us/step
Epoch 47/50
23/23 - 0s - loss: 0.6671 - 16ms/epoch - 705us/step
Epoch 48/50
23/23 - 0s - loss: 0.7004 - 19ms/epoch - 826us/step
Epoch 49/50
23/23 - 0s - loss: 0.4738 - 17ms/epoch - 736us/step
Epoch 50/50
23/23 - 0s - loss: 0.5142 - 16ms/epoch - 714us/step
10/10 [==============================] - 0s 800us/step
Epoch 1/50
23/23 - 0s - loss: 0.4983 - 25ms/epoch - 1ms/step
Epoch 2/50
23/23 - 0s - loss: 0.6596 - 19ms/epoch - 828us/step
Epoch 3/50
23/23 - 0s - loss: 0.5223 - 17ms/epoch - 727us/step
Epoch 4/50
23/23 - 0s - loss: 0.5716 - 16ms/epoch - 701us/step
Epoch 5/50
23/23 - 0s - loss: 0.5633 - 18ms/epoch - 784us/step
Epoch 6/50
23/23 - 0s - loss: 0.7350 - 17ms/epoch - 749us/step
Epoch 7/50
23/23 - 0s - loss: 0.4443 - 17ms/epoch - 728us

23/23 - 0s - loss: 0.4266 - 16ms/epoch - 716us/step
Epoch 20/50
23/23 - 0s - loss: 0.5297 - 17ms/epoch - 749us/step
Epoch 21/50
23/23 - 0s - loss: 0.6122 - 17ms/epoch - 727us/step
Epoch 22/50
23/23 - 0s - loss: 0.4344 - 17ms/epoch - 742us/step
Epoch 23/50
23/23 - 0s - loss: 0.4434 - 17ms/epoch - 751us/step
Epoch 24/50
23/23 - 0s - loss: 0.5168 - 17ms/epoch - 730us/step
Epoch 25/50
23/23 - 0s - loss: 0.4847 - 16ms/epoch - 697us/step
Epoch 26/50
23/23 - 0s - loss: 0.4451 - 20ms/epoch - 853us/step
Epoch 27/50
23/23 - 0s - loss: 0.4264 - 17ms/epoch - 719us/step
Epoch 28/50
23/23 - 0s - loss: 0.4608 - 16ms/epoch - 707us/step
Epoch 29/50
23/23 - 0s - loss: 0.5574 - 17ms/epoch - 734us/step
Epoch 30/50
23/23 - 0s - loss: 0.4184 - 17ms/epoch - 740us/step
Epoch 31/50
23/23 - 0s - loss: 0.5209 - 18ms/epoch - 802us/step
Epoch 32/50
23/23 - 0s - loss: 0.5139 - 19ms/epoch - 844us/step
Epoch 33/50
23/23 - 0s - loss: 0.4818 - 18ms/epoch - 789us/step
Epoch 34/50
23/23 - 0s - loss: 0.4214 - 18ms/epoch -

Epoch 46/50
23/23 - 0s - loss: 0.5311 - 23ms/epoch - 989us/step
Epoch 47/50
23/23 - 0s - loss: 0.4883 - 18ms/epoch - 801us/step
Epoch 48/50
23/23 - 0s - loss: 0.4757 - 18ms/epoch - 776us/step
Epoch 49/50
23/23 - 0s - loss: 0.4806 - 17ms/epoch - 752us/step
Epoch 50/50
23/23 - 0s - loss: 0.5385 - 23ms/epoch - 987us/step
10/10 [==============================] - 0s 895us/step
Epoch 1/50
23/23 - 0s - loss: 0.5808 - 23ms/epoch - 991us/step
Epoch 2/50
23/23 - 0s - loss: 0.5553 - 17ms/epoch - 719us/step
Epoch 3/50
23/23 - 0s - loss: 0.5811 - 17ms/epoch - 743us/step
Epoch 4/50
23/23 - 0s - loss: 0.6422 - 16ms/epoch - 693us/step
Epoch 5/50
23/23 - 0s - loss: 0.4761 - 18ms/epoch - 779us/step
Epoch 6/50
23/23 - 0s - loss: 0.4724 - 20ms/epoch - 849us/step
Epoch 7/50
23/23 - 0s - loss: 0.4304 - 18ms/epoch - 772us/step
Epoch 8/50
23/23 - 0s - loss: 0.5358 - 17ms/epoch - 737us/step
Epoch 9/50
23/23 - 0s - loss: 0.4262 - 22ms/epoch - 962us/step
Epoch 10/50
23/23 - 0s - loss: 0.4594 - 17ms/epoch - 750us

Epoch 22/50
23/23 - 0s - loss: 0.4528 - 16ms/epoch - 714us/step
Epoch 23/50
23/23 - 0s - loss: 0.4651 - 17ms/epoch - 719us/step
Epoch 24/50
23/23 - 0s - loss: 0.4218 - 17ms/epoch - 757us/step
Epoch 25/50
23/23 - 0s - loss: 0.4111 - 17ms/epoch - 736us/step
Epoch 26/50
23/23 - 0s - loss: 0.4453 - 17ms/epoch - 726us/step
Epoch 27/50
23/23 - 0s - loss: 0.4058 - 18ms/epoch - 770us/step
Epoch 28/50
23/23 - 0s - loss: 0.4701 - 17ms/epoch - 755us/step
Epoch 29/50
23/23 - 0s - loss: 0.9239 - 17ms/epoch - 746us/step
Epoch 30/50
23/23 - 0s - loss: 0.5929 - 18ms/epoch - 765us/step
Epoch 31/50
23/23 - 0s - loss: 0.4619 - 17ms/epoch - 750us/step
Epoch 32/50
23/23 - 0s - loss: 0.4470 - 17ms/epoch - 750us/step
Epoch 33/50
23/23 - 0s - loss: 0.4204 - 17ms/epoch - 754us/step
Epoch 34/50
23/23 - 0s - loss: 0.4426 - 17ms/epoch - 721us/step
Epoch 35/50
23/23 - 0s - loss: 0.4621 - 17ms/epoch - 730us/step
Epoch 36/50
23/23 - 0s - loss: 0.4125 - 17ms/epoch - 741us/step
Epoch 37/50
23/23 - 0s - loss: 0.5190 - 

23/23 - 0s - loss: 0.4937 - 18ms/epoch - 796us/step
Epoch 50/50
23/23 - 0s - loss: 0.4805 - 17ms/epoch - 729us/step
10/10 [==============================] - 0s 900us/step
Epoch 1/50
23/23 - 0s - loss: 0.4697 - 20ms/epoch - 853us/step
Epoch 2/50
23/23 - 0s - loss: 0.5514 - 17ms/epoch - 730us/step
Epoch 3/50
23/23 - 0s - loss: 0.5102 - 17ms/epoch - 728us/step
Epoch 4/50
23/23 - 0s - loss: 0.4745 - 17ms/epoch - 735us/step
Epoch 5/50
23/23 - 0s - loss: 0.4417 - 17ms/epoch - 719us/step
Epoch 6/50
23/23 - 0s - loss: 0.4309 - 16ms/epoch - 714us/step
Epoch 7/50
23/23 - 0s - loss: 0.4196 - 16ms/epoch - 716us/step
Epoch 8/50
23/23 - 0s - loss: 0.5305 - 16ms/epoch - 717us/step
Epoch 9/50
23/23 - 0s - loss: 0.5875 - 17ms/epoch - 723us/step
Epoch 10/50
23/23 - 0s - loss: 0.5656 - 17ms/epoch - 718us/step
Epoch 11/50
23/23 - 0s - loss: 0.4234 - 17ms/epoch - 727us/step
Epoch 12/50
23/23 - 0s - loss: 0.4474 - 17ms/epoch - 733us/step
Epoch 13/50
23/23 - 0s - loss: 0.4308 - 17ms/epoch - 729us/step
Epoch 

Epoch 26/50
23/23 - 0s - loss: 0.4415 - 17ms/epoch - 749us/step
Epoch 27/50
23/23 - 0s - loss: 0.5998 - 16ms/epoch - 714us/step
Epoch 28/50
23/23 - 0s - loss: 0.5006 - 17ms/epoch - 740us/step
Epoch 29/50
23/23 - 0s - loss: 0.4493 - 16ms/epoch - 681us/step
Epoch 30/50
23/23 - 0s - loss: 0.5050 - 19ms/epoch - 808us/step
Epoch 31/50
23/23 - 0s - loss: 0.4971 - 16ms/epoch - 708us/step
Epoch 32/50
23/23 - 0s - loss: 0.5810 - 16ms/epoch - 697us/step
Epoch 33/50
23/23 - 0s - loss: 0.4966 - 16ms/epoch - 693us/step
Epoch 34/50
23/23 - 0s - loss: 0.4190 - 17ms/epoch - 739us/step
Epoch 35/50
23/23 - 0s - loss: 0.5815 - 18ms/epoch - 772us/step
Epoch 36/50
23/23 - 0s - loss: 0.5625 - 24ms/epoch - 1ms/step
Epoch 37/50
23/23 - 0s - loss: 0.4743 - 31ms/epoch - 1ms/step
Epoch 38/50
23/23 - 0s - loss: 0.5247 - 26ms/epoch - 1ms/step
Epoch 39/50
23/23 - 0s - loss: 0.4905 - 24ms/epoch - 1ms/step
Epoch 40/50
23/23 - 0s - loss: 0.4377 - 32ms/epoch - 1ms/step
Epoch 41/50
23/23 - 0s - loss: 0.4616 - 25ms/epoch

Epoch 2/50
23/23 - 0s - loss: 0.5319 - 16ms/epoch - 681us/step
Epoch 3/50
23/23 - 0s - loss: 0.4174 - 16ms/epoch - 700us/step
Epoch 4/50
23/23 - 0s - loss: 0.4748 - 16ms/epoch - 678us/step
Epoch 5/50
23/23 - 0s - loss: 0.4193 - 16ms/epoch - 680us/step
Epoch 6/50
23/23 - 0s - loss: 0.4352 - 21ms/epoch - 895us/step
Epoch 7/50
23/23 - 0s - loss: 0.4642 - 16ms/epoch - 704us/step
Epoch 8/50
23/23 - 0s - loss: 0.4022 - 16ms/epoch - 686us/step
Epoch 9/50
23/23 - 0s - loss: 0.5109 - 16ms/epoch - 696us/step
Epoch 10/50
23/23 - 0s - loss: 0.4674 - 16ms/epoch - 707us/step
Epoch 11/50
23/23 - 0s - loss: 0.6204 - 16ms/epoch - 712us/step
Epoch 12/50
23/23 - 0s - loss: 0.5305 - 16ms/epoch - 689us/step
Epoch 13/50
23/23 - 0s - loss: 0.4015 - 17ms/epoch - 744us/step
Epoch 14/50
23/23 - 0s - loss: 0.4154 - 16ms/epoch - 679us/step
Epoch 15/50
23/23 - 0s - loss: 0.4105 - 15ms/epoch - 672us/step
Epoch 16/50
23/23 - 0s - loss: 0.4553 - 16ms/epoch - 684us/step
Epoch 17/50
23/23 - 0s - loss: 0.5092 - 16ms/epo

Epoch 29/50
23/23 - 0s - loss: 0.4071 - 20ms/epoch - 888us/step
Epoch 30/50
23/23 - 0s - loss: 0.4658 - 17ms/epoch - 747us/step
Epoch 31/50
23/23 - 0s - loss: 0.4531 - 19ms/epoch - 814us/step
Epoch 32/50
23/23 - 0s - loss: 0.7152 - 20ms/epoch - 852us/step
Epoch 33/50
23/23 - 0s - loss: 0.4890 - 17ms/epoch - 736us/step
Epoch 34/50
23/23 - 0s - loss: 0.4501 - 17ms/epoch - 739us/step
Epoch 35/50
23/23 - 0s - loss: 0.4264 - 17ms/epoch - 745us/step
Epoch 36/50
23/23 - 0s - loss: 0.4415 - 17ms/epoch - 755us/step
Epoch 37/50
23/23 - 0s - loss: 0.4215 - 17ms/epoch - 726us/step
Epoch 38/50
23/23 - 0s - loss: 0.5285 - 16ms/epoch - 715us/step
Epoch 39/50
23/23 - 0s - loss: 0.4514 - 17ms/epoch - 725us/step
Epoch 40/50
23/23 - 0s - loss: 0.4668 - 16ms/epoch - 713us/step
Epoch 41/50
23/23 - 0s - loss: 0.4256 - 17ms/epoch - 745us/step
Epoch 42/50
23/23 - 0s - loss: 0.4393 - 17ms/epoch - 736us/step
Epoch 43/50
23/23 - 0s - loss: 0.4901 - 16ms/epoch - 709us/step
Epoch 44/50
23/23 - 0s - loss: 0.4677 - 

23/23 - 0s - loss: 0.4697 - 23ms/epoch - 1ms/step
Epoch 6/50
23/23 - 0s - loss: 0.5238 - 19ms/epoch - 835us/step
Epoch 7/50
23/23 - 0s - loss: 0.5514 - 16ms/epoch - 699us/step
Epoch 8/50
23/23 - 0s - loss: 0.4481 - 16ms/epoch - 685us/step
Epoch 9/50
23/23 - 0s - loss: 0.3925 - 17ms/epoch - 735us/step
Epoch 10/50
23/23 - 0s - loss: 0.4962 - 15ms/epoch - 673us/step
Epoch 11/50
23/23 - 0s - loss: 0.4514 - 16ms/epoch - 683us/step
Epoch 12/50
23/23 - 0s - loss: 0.4416 - 16ms/epoch - 687us/step
Epoch 13/50
23/23 - 0s - loss: 0.5109 - 15ms/epoch - 672us/step
Epoch 14/50
23/23 - 0s - loss: 0.4331 - 16ms/epoch - 674us/step
Epoch 15/50
23/23 - 0s - loss: 0.4393 - 18ms/epoch - 766us/step
Epoch 16/50
23/23 - 0s - loss: 0.5945 - 18ms/epoch - 788us/step
Epoch 17/50
23/23 - 0s - loss: 0.4771 - 17ms/epoch - 729us/step
Epoch 18/50
23/23 - 0s - loss: 0.4159 - 16ms/epoch - 680us/step
Epoch 19/50
23/23 - 0s - loss: 0.4084 - 19ms/epoch - 822us/step
Epoch 20/50
23/23 - 0s - loss: 0.6216 - 17ms/epoch - 753us

### Comparison to the mean of the MSEs compred to which from A

In [4]:
print("The final MSE at the 50th time is {}.".format(MSE_list[-1]))
print("The mean of the MSE is {}.".format(np.mean(MSE_list)))
print("The standard deviation of the MSE is {}".format(np.std(MSE_list)))


The final MSE at the 50th time is 1.5316254715318072.
The mean of the MSE is 0.7051157050923686.
The standard deviation of the MSE is 1.1030832306031149


## How does the mean of the MSEs compare to which from Step A?

#### Mean of MSEs from Step A: 56.564149782825716
#### Mean of MSEs from Step B: 0.7051157050923686
### Thus, the mean of the MSEs is decreased compare to which from Step A.